code/game/machinery/vending/cm_vending.dm

```diff
@@ -902,108 +873,190 @@ IN_USE						used for vending/denying
-			prod_available = TRUE
-		else if(p_amount == 0)
-			p_name += ": SOLD OUT"				//Negative numbers (-1) used for categories.
+		var/prod_available = p_amount > 0		//checking if it's available
+		var/list/initial_vals = initial_product_count
+
+		var/result = product_icon_list[p_name]
+		var/initial_amount = initial_vals[p_name]
+		var/is_category = p_amount < 0
+		var/typepath = myprod[3]
+		var/obj/item/I = typepath
 
 		//forming new list with index, name, amount, available or not, color and add it to display_list
-		display_list += list(list("prod_index" = i, "prod_name" = p_name, "prod_amount" = p_amount, "prod_available" = prod_available, "prod_color" = myprod[4]))
+		var/display_item = list(
+			"prod_index" = i,
+			"prod_name" = p_name,
+			"prod_available" = prod_available,
+			"prod_color" = myprod[4],
+			"prod_initial" = initial_amount,
+			"prod_icon" = result,
+			"prod_desc" = initial(I.desc)
+		)
+
+		if (is_category == 1)
+			ui_categories += list(list(
+				"name" = p_name,
+				"items" = list()
+			))
+			continue
+
+		if (!LAZYLEN(ui_categories))
+			ui_categories += list(list(
+				"name" = "",
+				"items" = list()
+			))
+		var/last_index = LAZYLEN(ui_categories)
+		var/last_category = ui_categories[last_index]
+		last_category["items"] += list(display_item)
+	data["displayed_categories"] = ui_categories
+	return data
+
+/obj/structure/machinery/cm_vending/sorted/ui_data(mob/user)
+	var/list/data = list()
 
+	var/list/ui_listed_products = get_listed_products(user)
+	if(!LAZYLEN(ui_listed_products))	//runtimed for vendors without goods in them
+		to_chat(user, SPAN_WARNING("Vendor wasn't properly initialized, tell an admin!"))
+		return
 
-	var/list/data = list(
-		"vendor_name" = name,
-		"theme" = vendor_theme,
-		"displayed_records" = display_list,
-	)
+	var/list/ui_categories = list()
 
-	ui = nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+	for (var/i in 1 to length(ui_listed_products))
+		var/list/myprod = ui_listed_products[i]	//we take one list from listed_products
+		var/p_amount = myprod[2]				//amount left
+		ui_categories += list(p_amount)
+	data["stock_listing"] = ui_categories
+	return data
 
+/obj/structure/machinery/cm_vending/sorted/tgui_interact(mob/user, datum/tgui/ui)
+	if(!ishuman(user))
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
 	if (!ui)
-		ui = new(user, src, ui_key, "cm_vending_sorted.tmpl", name , 600, 700)
-		ui.set_initial_data(data)
+		ui = new(user, src, "VendingSorted", name)
 		ui.open()
-		ui.set_auto_update(0)		//here we need autoupdate because items can be vended by others and are limited
 
-/obj/structure/machinery/cm_vending/sorted/Topic(href, href_list)
+/obj/structure/machinery/cm_vending/sorted/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	. = ..()
+	if(.)
+		return
+	if(!(in_range(src, usr) && isturf(loc) && ishuman(usr)))
+		return
```

Comment:
should already be handled by ui_state and ui_status

Reply:
ok, I'm not entirely sure on how all of the backend works - thanks for the info!